### PR TITLE
First working  version of netconf adapter.

### PR DESCRIPTION
**Pull request by sjames-iberis**

### CODE REVIEW

File: cmd/server.go
Comment (SeanCondon) [inline cmd/server.go:61]: we generally use 5150 as the default port (yes after the Van Halen album, but which was named after a California law)

File: cmd/server.go
Comment (SeanCondon) [inline cmd/server.go:64]: probably should not have a default here

File: cmd/server.go
Comment (SeanCondon) [inline cmd/server.go:97]: maybe feedback the port number

File: cmd/version.go
Comment (SeanCondon) [inline cmd/version.go:42]: I like this - we should adopt something like this in our modules

File: get.all.gnmi
Comment (SeanCondon) [inline get.all.gnmi:1]: I really like these - I will start using the same style for `onos-config`

File: gnmi_cli/README.md
Comment (SeanCondon) [inline gnmi_cli/README.md:7]: maybe address should be `gnmi-netconf-adapter:5150` - as that should be the service name in k8s

File: pkg/adapter/modeldata/gostruct/gen.go
Comment (SeanCondon) [inline pkg/adapter/modeldata/gostruct/gen.go:17]: If you use `go generate` from the Makefile it will create the "generated.go" file and it will not have to be versioned below.
I know there is an issue with the version of ygot, but this should be dealt with through the `go.mod`.

Also I should point out that the Docker base image we use "onosproject/golang-build:stable" has the "generator" from ygot in it - I think this is the right place to adjust the "ygot" version

**Review by SeanCondon — COMMENTED** [review]
I got partly through it - set.go is next - I will resume in the morning

File: pkg/adapter/capabilities.go
Comment (SeanCondon) [inline pkg/adapter/capabilities.go]: year can be 2020. Same elsewhere

File: pkg/adapter/adapter_test.go
Comment (SeanCondon) [inline pkg/adapter/adapter_test.go]: this could be broken out in to a separate file `capabilities_test.go` - would help make the files more manageable in size

File: pkg/adapter/adapter_test.go
Comment (SeanCondon) [inline pkg/adapter/adapter_test.go]: this could be broken out in to a separate file `get_test.go` - would help make the files more manageable in size

File: pkg/adapter/adapter_test.go
Comment (SeanCondon) [inline pkg/adapter/adapter_test.go]: this could be broken out in to a separate file `subscribe_test.go` - would help make the files more manageable in size

File: pkg/adapter/adapter_test.go
Comment (SeanCondon) [inline pkg/adapter/adapter_test.go]: this file might be called `test_set.go` - with other test files corresponding to `subscribe.go`, `get.go` etc. It would help make the files more manageable in size

File: pkg/adapter/modeldata/modeldata.go
Comment (SeanCondon) [inline pkg/adapter/modeldata/modeldata.go:20]: I think the whole model plugin here could be loaded in to `gnmi_netconf_adapter` like the model plugins are now loaded in to `onos-config` i.e. as a sidecar container in Kubernetes. I think this is a good way that the adapter could reuse and be kept in sync with the model plugins that `onos-config` uses. The sidecar loads and copies its `'so'` file over to `onos-config` which loads it as a shared library.

File: pkg/adapter/adapter_test.go
Comment (SeanCondon) [inline pkg/adapter/adapter_test.go]: i find `pb` a particularly meaningless and unintuitive alias for `gnmi` - I'd rather just see 'gnmi` used.

File: pkg/adapter/adapter_test.go
Comment (SeanCondon) [inline pkg/adapter/adapter_test.go]: could use `assert.NoError(t, err, "error in creating server: %v", err)` here instead

File: pkg/adapter/adapter_test.go
Comment (SeanCondon) [inline pkg/adapter/adapter_test.go]: same here

File: pkg/adapter/adapter_test.go
Comment (SeanCondon) [inline pkg/adapter/adapter_test.go]: Could use `assert.Equal(t, tc.wantRetCode, gotRetStatus.Code(), "Unexpected return code")` here instead. The values will get printed in the error.

File: pkg/adapter/modeldata/modeldata.go
Comment (SeanCondon) [reply pkg/adapter/modeldata/modeldata.go:20]: For the Unit tests you could test with `TestDevice-1.0.0` which have a very wide range of YANG constructs. You could depend on this like
```
	td1 "github.com/onosproject/config-models/modelplugin/testdevice-1.0.0/testdevice_1_0_0"
	td2 "github.com/onosproject/config-models/modelplugin/testdevice-2.0.0/testdevice_2_0_0"
```

**Review by SeanCondon — APPROVED** [review]
This is really great work.

My comments are just suggestions - I'm OK to merge this as is, and if comments are to be addressed it can be done later (depending on the disposition)

File: pkg/adapter/capabilities.go
Comment (sjames-iberis) [reply pkg/adapter/capabilities.go]: Done.

File: pkg/adapter/adapter_test.go
Comment (sjames-iberis) [reply pkg/adapter/adapter_test.go]: Done,

File: pkg/adapter/adapter_test.go
Comment (sjames-iberis) [reply pkg/adapter/adapter_test.go]: Done.

File: pkg/adapter/adapter_test.go
Comment (sjames-iberis) [reply pkg/adapter/adapter_test.go]: Done

File: pkg/adapter/adapter_test.go
Comment (sjames-iberis) [reply pkg/adapter/adapter_test.go]: Done

File: pkg/adapter/adapter_test.go
Comment (sjames-iberis) [reply pkg/adapter/adapter_test.go]: Done

File: pkg/adapter/adapter_test.go
Comment (sjames-iberis) [reply pkg/adapter/adapter_test.go]: Done, after renaming adapter package from gnmi to adapter.

File: pkg/adapter/adapter_test.go
Comment (sjames-iberis) [reply pkg/adapter/adapter_test.go]: Done

File: pkg/adapter/adapter_test.go
Comment (sjames-iberis) [reply pkg/adapter/adapter_test.go]: Done.

File: pkg/adapter/modeldata/modeldata.go
Comment (sjames-iberis) [reply pkg/adapter/modeldata/modeldata.go:20]: Regarding the test models, I would prefer to hold off on this for the time being.
I agree that the test yang and generated model  is a useful resource; it certainly covers more than the junos fragment that we are currently using.
However, I can see that it would expose some of the known limitations in the current adapter implementation, relating to data type transformations; these are marked by TODOs in the code.
The TODOs will obviously need to be addressed in due course, but:
	•	For the time being I don’t think they compromise the value of the initial test/demo
	•	When they are implemented, I’d prefer to have access to a real schema/device combination to inform and validate the implementation

**Comment by SeanCondon** [comment]
@sjames-iberis I don't know if you saw the first set of comments I had left above e.g. using the port number 5150 etc. Anyway I want to merge this now and move on - if you're going to address these comments it can be done in a future patch

**Comment by sjames-iberis** [comment]
@SeanCondon - I had been focused on the adapter package.  
I had seen and was not ignoring the comments on the cmd package and was intending to discuss with Kieran. Will follow up next week.  
Thanks for the feedback.

**Pull request merged**
